Initial version of the psp-user-group policy (#1)
kubewarden/psp-user-group: README.md

This Kubewarden Policy is a replacement for the Kubernetes Pod Security 
Policy that controls containers [user and groups](https://kubernetes.io/docs/concepts/policy/pod-security-policy/#users-and-groups).

> This policy is used to control the user and groups user in containers.

Reviewer: ```suggestion
This policy is used to control users and groups in containers.
```

## Settings

The policy has three settings:

> * `run_as_user`: Controls which user ID the containers are run with. It is a JSON object with two attributes: `rule` and `ranges`. The `rule` allowed values are:

Reviewer: ```suggestion
* `run_as_user`: Controls which user ID the containers are run with.
```

Given that is explained later on.

* `run_as_group`:  Controls which primary group ID the containers are run with.
* `supplemental_groups`: Controls which group IDs containers add.

> All the three settings are JSON objects composed by two attributes: `rule` and `ranges`. The `rule` attribute define

Reviewer: ```suggestion
All the three settings are JSON objects composed by two attributes: `rule` and `ranges`. The `rule` attribute defines
```

> the strategy used by the policy to enforce user and groups used in containers. The available strategies are:

Reviewer: ```suggestion
the strategy used by the policy to enforce users and groups used in containers. The available strategies are:
```

	* `MayRunAs` - Does not require that `RunAsGroup` be specified. However, when `RunAsGroup` is specified, they have to fall in the defined range.
	* `RunAsAny` - No default provided. Allows any `runAsGroup` to be specified.
* `supplemental_groups`: 
	* `MustRunAs` - Requires at least one range to be specified. Uses the minimum value of the first range as the default. Validates against all ranges.
	* `MayRunAs` - Requires at least one range to be specified. Allows `supplementalGroups` to be left unset without providing a default. Validates against all ranges if `supplementalGroups` is set.
	* `RunAsAny` - No default provided. Allows any `supplementalGroups` to be specified

The `ranges` is a list of JSON objects with two attributes: `min` and `max`. Each range object define the user/group ID range used by the rule. 

### Examples

To enforce that user and groups must be set and it should be in the defined ranges:

```json
{
        "run_as_user": {
                "rule": "MustRunAs",
                "ranges": [
                        {"min": 1000, "max": 2000},

>                         {"min": 2001, "max": 3000}

Reviewer: In the example I would add a gap, so it's clear the usage. For example:

```
                        {"min": 1000, "max": 1999},
                        {"min": 3000, "max": 3999}
```

In this example, 2000-2999 is out of any allowed ranges, and would be rejected.

                ]
        },
        "run_as_group": {
                "rule": "MustRunAs",
                "ranges": [
                        {"min": 1000, "max": 2000},

>                         {"min": 2001, "max": 3000}

Reviewer: I would add a gap here as well.

                ]
        },
        "supplemental_groups":{
                "rule": "MustRunAs",
                "ranges": [
                        {"min": 1000, "max": 2000},

>                         {"min": 2001, "max": 3000}

Reviewer: I would add a gap here as well.

                ]
        }
}
```

To allow any user and group:

```json
{
	"run_as_user": {
    		"rule": "RunAsAny"
	},
	"run_as_group": {
      		"rule": "RunAsAny"
	},
    	"supplemental_groups":{
      		"rule": "RunAsAny"
    	}

> }

Reviewer: It seems there is a mix of spaces and tabs in this example.

```

> To force run container with non root user but any group:

Reviewer: ```suggestion
To force running the container with non root user but any group:
```

```json
{
	"run_as_user": {
    		"rule": "MustRunAsNonRoot"
	},
	"run_as_group": {
      		"rule": "RunAsAny"
	},
    	"supplemental_groups":{
      		"rule": "RunAsAny"

>     	}

Reviewer: Mix of tabs and spaces here too.

}
```

To enforce a group when the container has some group defined

```json
{
	"run_as_user": {
    		"rule": "RunAsAny"
	},
	"run_as_group": {
    		"rule": "MayRunAs",
		"ranges": [
			{"min": 1000, "max": 2000},
			{"min": 2001, "max": 3000}
		]
	},
    	"supplemental_groups":{
    		"rule": "MayRunAs",
		"ranges": [
			{"min": 1000, "max": 2000},
			{"min": 2001, "max": 3000}
		]

>     	}

Reviewer: Mix of tabs and spaces. I think in several places of this file, not only here.
